### ISSUE TITLE
Access to constraint references

### ISSUE THREAD

**Issue by bcdandurand** [lanl-ansi/PowerModels.jl#661]
In our application, it is required to be able to reference constraints generated through the PowerModels.jl interface, either to modify them or to access dual values.

My current fork is up-to-date with the lanl-ansi master branch except for a few changes to have some of the methods that construct constraints to return the constraint references:

https://github.com/bcdandurand/PowerModels.jl

If I missed some existing feature that would have made these modifications unnecessary, please let me know. Otherwise, may I ask that a future release of PowerModels provide a mechanism to give the user access to constraint references from generated constraints? Thank you.

**Comment by ccoffrin** [comment]
The datastore `pm.con` us where the constraint references live.  However, we are not currently very consistent about which ones are added there and what their names are.  At the moment it is just the power balance and branch limits needed for some LMP analysis.

**Comment by bcdandurand** [comment]
For now then, we need to use our fork until the PowerModels master version allows access to the needed constraint references. We shall keep our fork updated with the PowerModels master version until then.

**Comment by bcdandurand** [comment]
After additional experience using PowerModels, it might be better that standard constraint names be established in a future version, rather than having the constraint generating functions return references. That way, if the user generates a model first, then later attaches a solver, the user can otherwise obtain the reference via the constraint name.

**Comment by ccoffrin** [comment]
Agreed, the convention would that constraint references are stored in `pm.con` under a standard naming scheme.  For the moment this is only done for a few constraint cases (power balance and branch flow) and not well tested across different formulations.

Adding this feature will incur a non trivial amount of code maintenance and also put a bigger burden on PowerModels extension implementation.  So I am hesitant to commit to it until I have a sufficient body of motivating use cases.

**Comment by odow** [comment]
We discussed this, and for now we're going to close this as out-of-scope.

The main talking points were:

 * Yes, it's a shame that PowerModels doesn't have a standardized way to access all constraints
 * But adopting such a scheme is too much of an effort and introduces a large maintenance and documentation burden for a relatively small number of users
 * There are two main reason for wanting this:
   1. You want the dual variable of a particular constraint (#841), (#876)
   2. You want some sort of decomposition (#942)
 * In the first case, we could consider a small number of requests to add the dual for practically relevant constraints (e.g., energy balances for prices)
 * In the second case, PowerModels.jl is not the right tool for the job. You probably want instead to use the data parsing features of PowerModels, and then build your own model that you control the data structure of. See https://github.com/lanl-ansi/PowerModelsAnnex.jl for examples.
